File: web/packages/client/src/project-content.ts
import { ProtocolError } from "./errors.js";

export const PROJECT_CONTENT_OPERATIONS = ["list", "read", "body", "create", "revise", "deprecate", "evidence", "evidence_list", "attach", "attachments", "outcome_list", "outcome_read", "outcome_write"] as const;
export type ProjectContentOperation = typeof PROJECT_CONTENT_OPERATIONS[number];
export type ProjectContentInput = Readonly<Record<string, unknown>>;
export type ProjectContentOutput = Readonly<Record<string, unknown>>;
export type ProjectContentRequest = Readonly<{ operation: ProjectContentOperation; input: ProjectContentInput }>;
export type ProjectContentResult = Readonly<{ operation: ProjectContentOperation; output: ProjectContentOutput }>;

export function projectContentOperation(value: unknown): ProjectContentOperation {
  if (typeof value !== "string" || !(PROJECT_CONTENT_OPERATIONS as readonly string[]).includes(value)) throw new ProtocolError("malformed");
  return value as ProjectContentOperation;
}

/** Content uses bounded JSON numbers; refuse precision loss at the boundary. */
export function projectContentObject(value: unknown): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) throw new ProtocolError("malformed");
  const visit = (item: unknown, depth: number): void => {
    if (depth > 16 || typeof item === "number" && (!Number.isSafeInteger(item) || item < 0)) throw new ProtocolError("malformed");
    if (Array.isArray(item)) { if (item.length > 64) throw new ProtocolError("malformed"); item.forEach((child) => visit(child, depth + 1)); }
    else if (item !== null && typeof item === "object") Object.values(item).forEach((child) => visit(child, depth + 1));
    else if (typeof item !== "string" && typeof item !== "boolean" && typeof item !== "number" && item !== null) throw new ProtocolError("malformed");
  };
  visit(value, 0);
  return value as Record<string, unknown>;
}

File: internal/browserprotocol/project_content.go
package browserprotocol

import (
	"bytes"
	"encoding/json"
)

// ProjectContent is one bounded, lazy project-library operation. Keeping one
// finite operation union avoids adding a service or exposing API credentials.
type ProjectContent struct {
	Operation string          `json:"operation"`
	Input     json.RawMessage `json:"input"`
}

type ProjectContentResult struct {
	Operation string          `json:"operation"`
	Output    json.RawMessage `json:"output"`
}

func EncodeProjectContent(id string, value ProjectContent) ([]byte, error) {
	return encodeControl(TypeProjectContent, id, value)
}
func EncodeProjectContentResult(id string, value ProjectContentResult) ([]byte, error) {
	return encodeControl(TypeProjectContentResult, id, value)
}

func validProjectContent(kind MessageType, body any) error {
	var operation string
	var raw json.RawMessage
	if kind == TypeProjectContent {
		value, ok := indirect(body).(ProjectContent)
		if !ok {
			return ErrMalformed
		}
		operation, raw = value.Operation, value.Input
	} else {
		value, ok := indirect(body).(ProjectContentResult)
		if !ok {
			return ErrMalformed
		}
		operation, raw = value.Operation, value.Output
	}
	if !validProjectOperation(operation) {
		return ErrMalformed
	}
	if len(raw) > MaxControlBytes-1024 {
		return ErrOversized
	}
	trimmed := bytes.TrimSpace(raw)
	if len(trimmed) == 0 || trimmed[0] != '{' || !json.Valid(raw) {
		return ErrMalformed
	}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.UseNumber()
	// Use the same integer, depth and collection bounds as the complete wire
	// decoder. Unrepresentable stored metadata is an explicit bounded error.
	if err := scanJSONValue(decoder, 2, MaxArrayItems); err != nil {
		if kind == TypeProjectContentResult {
			return ErrOversized
		}
		return ErrMalformed
	}
	return nil
}

func validProjectOperation(operation string) bool {
	switch operation {
	case "list", "read", "body", "create", "revise", "deprecate", "evidence", "evidence_list", "attach", "attachments", "outcome_list", "outcome_read", "outcome_write":
		return true
	default:
		return false
	}
}
